irohad/execution/impl/command_executor.cpp

```diff
@@ -488,220 +490,481 @@ namespace iroha {
     // asset types, now: anyone having permission "can_add_asset_qty" can add
     // any asset
-    return creator_account_id == command.accountId()
-        and checkAccountRolePermission(
-                creator_account_id, queries, Role::kAddAssetQty);
+    if (creator_account_id != command.accountId()) {
+      return makeCommandError(
+          "has permission command validation failed: creator account "
+              + creator_account_id + " is not command account "
+              + command.accountId(),
+          command_name);
+    }
+    if (not checkAccountRolePermission(
+            creator_account_id, queries, Role::kAddAssetQty)) {
+      return makeCommandError(
+          "has permission command validation failed: account "
+              + creator_account_id + " does not have kAddAssetQty",
+          command_name);
+    }
+    return {};
   }
 
-  bool CommandValidator::hasPermissions(
+  CommandResult CommandValidator::hasPermissions(
       const shared_model::interface::AddPeer &command,
       ametsuchi::WsvQuery &queries,
       const shared_model::interface::types::AccountIdType &creator_account_id) {
-    return checkAccountRolePermission(
-        creator_account_id, queries, Role::kAddPeer);
+    auto command_name = "AddPeer";
+    if (not checkAccountRolePermission(
+            creator_account_id, queries, Role::kAddPeer)) {
+      return makeCommandError(
+          "has permission command validation failed: account "
+              + creator_account_id + " does not have kAddPeer",
+          command_name);
+    }
+    return {};
   }
 
-  bool CommandValidator::hasPermissions(
+  CommandResult CommandValidator::hasPermissions(
       const shared_model::interface::AddSignatory &command,
       ametsuchi::WsvQuery &queries,
       const shared_model::interface::types::AccountIdType &creator_account_id) {
-    return
-        // Case 1. When command creator wants to add signatory to their
-        // account and he has permission CanAddSignatory
-        (command.accountId() == creator_account_id
-         and checkAccountRolePermission(
-                 creator_account_id, queries, Role::kAddSignatory))
-        or
-        // Case 2. Creator has granted permission for it
-        (queries.hasAccountGrantablePermission(creator_account_id,
-                                               command.accountId(),
-                                               Grantable::kAddMySignatory));
+    auto command_name = "AddSignatory";
+
+    auto creator_adds_signatory_to_his_account = [&creator_account_id,
+                                                  &command]() {
+      return creator_account_id == command.accountId();
+    };
+    auto creator_has_permission_on_own = [&creator_account_id, &queries]() {
+      return checkAccountRolePermission(
+          creator_account_id, queries, Role::kAddSignatory);
+    };
+    auto creator_has_grantable_permission =
```

Comment:
How lazy is better than lambda here? I don't see where it helps to avoid redundant `queries.hasAccountGrantablePermission ` calls

Reply:
Doesn't it help? I call for ```creator_has_grantable_permission``` twice, so the result is memorised from the first execution of ```queries.hasAccountGrantablePermission```. Or I understood lazy here in a wrong way?